lib/invoker.js

```diff
@@ -24,21 +25,53 @@ const invoke = (userFunction) => {
     };
 };
 
-module.exports = (userFunctionUri, port) => {
-    const userFunction = lookUpFunction(userFunctionUri);
-
-    const server = new grpc.Server();
-    server.addService(services.RiffService, {invoke: invoke(userFunction)});
+const executeInit = async (initHook, timeout, server) => {
+    logger('Calling $init hook (if any)');
+    return await guardWithTimeout(initHook, timeout)
+        .then(() => logger('$init hook successfully completed (or skipped if none)'))
+        .catch((err) => {
+            logger(`Error during $init hook execution: ${err.toString()}`);
+            logger('Shutting down now');
+            server.forceShutdown();
+            return Promise.reject('$init error');
+        });
+};
 
+const registerShutdown = (globals, destroyHook, timeout, server) => {
     const shutdown = () => {
         logger('Graceful shutdown started');
-        server.tryShutdown(() => logger('Graceful shutdown completed'));
+        logger('Calling $destroy hook (if any)');
+
+        guardWithTimeout(destroyHook, timeout)
```

Comment:
the destroy hook should be called after the server has shutdown. Currently, the destroy will run while requests may be in flight.

Reply:
👌

Reply:
Done but I have not found a way to test that the shutdown is actually called after the server shutdown.